README.md
# scio-api

Web API for submitting and retrieving documents from Scio

## Build

```
$ lein deps
$ lein uberjar
```

## Usage

```bash
$ java -jar target/scio-api-0.1.0-SNAPSHOT-standalone.jar -h
Usage: java -jar scio-api-VERSION.jar [OPTION...]

-c, --config=CONFIGFILE      Specify ini location
-h, --help                   print this text

You can also override the default location of the .ini file by exporting the SCIOAPIINI environment variable.

report bugs to opensource@mnemonic.no
```

*examples*
```bash
$ java -jar scio-api-0.1.0-standalone.jar -c /opt/scio/scio.ini
```

```bash
$ SCIOAPIINI=/opt/scio/scio.ini java -jar scio-api-0.1.0-standalone.jar
```

## Configuration

*Consider sharing the .ini file between scio and scio-api*

```ini
[beanstalk]
queue = submit
host = localhost
port = 11300

[elasticsearch]
host = http://localhost:9201 http://localhost:9202

[storage]
storagedir = /tmp

[api]
port = 3000
```

### [beanstalk]

*queue*: This is the beanstalk tube shared by scio-api and scio. This is where scio retrieves the filenames
to scrape for indicators.

*host*: The hostname of the beanstalkd service

*port*: The ports running the beanstalkd service

### [elasticsearch]
*host*: A list of hosts separated by space. This is the list of *coordinators* in the case of a cluster.

### [storagedir]

*storagedir*: The directory where the submitted documents are stored. This can be the same place as scio stores the documents. *Important: scio must be able to read the files in this directory*

### [api]
*port*: The port the web API will be listening on.

## Tests using curl to verify everything works

### Submitting data

The Content-Type *MUST* be "application/json"
The data must be a json map with the keys "content" containing the file content as base64 and "filename" containing the basename of the file.

```bash
$ curl -H "Content-Type: application/json" -X POST --data '{"content": "MTI3LjAuMC4xCg==", "filename": "testfile.txt"}' http://localhost:3000/submit
ok⏎
$
```

### Downloading data

```bash
$ curl 'http://localhost:3000?id=77785fc7b151a325a39d2c40a7701cb57736f2ce34b7edbef4e538f42c1509d3'
{"filename":"testfile.txt","content":"MTI3LjAuMC4xCg==","encoding":"base64"}
$
```

## License

ISC License

Copyright (c) 2016-2018, mnemonic as <opensource@mnemonic.no>

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted, provided that the above
copyright notice and this permission notice appear in all copies.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE
OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
-

## [0.1.0] - 2019-04-25
### Changed
-

### Removed
-

### Fixed
-

### Added
-



test/scio_api/core_test.clj
(ns scio-api.core-test
  (:require [clojure.test :refer :all]
            [scio-api.core :refer :all]
            [ring.mock.request :as mock]))

(deftest test-upload
  (System/setProperty "*scio-api-ini*" "etc/sample.ini")
  (is (= {:status 200
          :bytes 4
          :body "ok"
          :headers {}}
         (api (-> (mock/request :post "/submit")
                  (mock/json-body {:filename "test.txt"
                                   :content "VGVzdA=="})))))
  (is (= {:status 200
          :bytes 4
          :body "ok"
          :headers {}}
         (api (-> (mock/request :post "/submit")
                  (mock/json-body {:filename "../../../../../../../../../../../../../../../root/evil.txt"
                                   :content "VGVzdA=="}))))))
 


project.clj
(defproject scio-api "0.1.0-SNAPSHOT"
  :description "API for uploading and downloading documents"
  :url "https://github.com/mnemonic-no/act-scio-api"
  :license {:name "ISC"
            :url "https://opensource.org/licenses/ISC"}
  :dependencies [[org.clojure/clojure "1.10.0"]
                 [org.clojure/tools.cli "0.4.2"]
                 [ring/ring-core "1.7.1"]
                 [ring/ring-json "0.4.0"]
                 [ring/ring-jetty-adapter "1.7.1"]
                 [ring/ring-mock "0.3.2"]
                 [clojure-ini "0.0.2"]
                 [beanstalk-clj "0.1.3"]
                 [cc.qbits/spandex "0.6.3"]
                 [compojure "1.6.1"]]
  :profiles {:uberjar {:aot :all
                       :main scio-api.core}}
  :main ^:skip-aot scio-api.core
  :plugins [[lein-ring "0.12.5"]]
  :ring {:handler scio-api.core/api}
  :repl-options {:init-ns scio-api.core})


src/scio_api/core.clj
(ns scio-api.core
  (:gen-class)
  (:require [compojure.core :refer :all]
            [compojure.route :as route]
            [scio-api.b64 :as b64]
            [qbits.spandex :as spandex]
            [clojure-ini.core :as clojure-ini]
            [clojure.java.io :refer [file output-stream input-stream copy]]
            [clojure.string :as str]
            [clojure.tools.cli :refer [parse-opts]]
            [beanstalk-clj.core :refer [with-beanstalkd beanstalkd-factory
                                        put use-tube]]
            [ring.adapter.jetty :refer :all]
            [ring.middleware.json :as ring-json]
            [ring.middleware.keyword-params :refer [wrap-keyword-params]]
            [ring.middleware.params :refer [wrap-params]]))


(if-let [ini-file (System/getenv "SCIOAPIINI")]
  (System/setProperty "*scio-api-ini*" ini-file)
  (System/setProperty "*scio-api-ini*" "/etc/scioapi.ini"))

(defn exit
  "Print message to stderr and exit with exit code"
  [msg exitcode]
  (.println *err* msg)
  (System/exit exitcode))

(defn slurp-bytes
  "Slurp the bytes from a slurpable thing"
  [x]
  (try
    (with-open [out (java.io.ByteArrayOutputStream.)]
      (copy (input-stream (file x)) out)
      (.toByteArray out))
    (catch java.io.FileNotFoundException e
      [])))

(defn get-config-file
  "read the config file path from the SCIOAPIINI environment path
  or default to /etc/scioapi.ini"
  []
  (let [cfg-file (System/getProperty "*scio-api-ini*")]
    (if (.isFile (file cfg-file))
      cfg-file
      (exit (str "\nCoult not find config file: "
                 cfg-file
                 "\nconsider to provide the SCIOAPIINI environment variable or use the '-c CONFIGFILE' argument.")
            1))))

(defn save-file
  ""
  [file-object content]
  (try (with-open [out (output-stream file-object)]
         (.write out content)
         {:error nil
          :filename (str file-object)
          :bytes (count content)})
       (catch Exception e {:error (.getMessage e)
                           :filename (str file-object)
                           :count 0})))

(defn register-submit
  ""
  [ini filename]
  (let [{:keys [host port queue]} (:beanstalk ini)]
    (with-beanstalkd (beanstalkd-factory host (Integer. port))
      (use-tube queue)
      (put filename))))

(defn handle-submit
  "Handle the submit api call. Write the content to the path specified in the
  storage sectuin in the .ini  file"
  [body]
  (let [_ (spit "/tmp/a" body)
        content (b64/decode (:content body))
        ini (clojure-ini/read-ini (get-config-file) :keywordize? true)
        file-name (.getName (file (:filename body))) ;; get basename to avoid directory traversal
        file-object (file (get-in ini [:storage :storagedir]) file-name)
        result (save-file file-object content)]
    (if (nil? (:error result))
      (do
        (register-submit ini (:filename result))
        {:status 200
         :bytes (:bytes result)
         :body "ok"})
      {:status 500
       :bytes 0
       :body (:error result)})))

(defn handle-download
  ""
  [id]
  (let [ini (clojure-ini/read-ini (get-config-file) :keywordize? true)
        host (get-in ini [:elasticsearch :host])
        hosts (str/split host #"\s")
        client (spandex/client {:hosts hosts})
        response (spandex/request client {:url [:scio :_search]
                                          :method :get
                                          :body {:query {:match {:_id id}}}})]
    (if-let [file-name (get-in response [:body :hits :hits 0 :_source :filename])]
      (let [file-base-name (.getName (file file-name))
            content (slurp-bytes (str (file (get-in ini [:storage :storagedir]) file-base-name)))]
        (if (seq content)
          {:status 200
           :bytes (count content)
           :body {:filename file-base-name
                  :content (b64/encode content)
                  :encoding "base64"}}
          {:status 404
           :bytes 0
           :body {:filename file-base-name
                  :content ""
                  :encoding "base64"}}))
      {:status 404
       :bytes 0
       :body {:filename ""
              :content ""
              :encoding "base64"}})))

(defroutes api-routes
  (POST "/submit" {:keys [params]} (handle-submit params))
  (GET "/download" [id] (handle-download id))
  (route/not-found {:body {:error "Page not found"}}))

(def api
  (-> api-routes
      wrap-keyword-params
      (ring-json/wrap-json-params {:keywords? true :bigdecimals? true})
      wrap-params
      (ring-json/wrap-json-response)))

(def cli-options
  "CLI Options"
  [["-c" "--config CONFIG" "Config File"
    :id :config
    :default "/etc/scio.ini"
    :validate [#(.isFile (file %)) "File not found!"]]
   ["-h" "--help"]])

(defn usage
  "Usage summary, Exit after call"
  []
  (do 
    (println "\nUsage: java -jar scio-api-VERSION.jar [OPTION...]\n")
    (println "-c, --config=CONFIGFILE      Specify ini location")
    (println "-h, --help                   print this text")
    (println "\nYou can also override the default location of the .ini file by exporting the SCIOAPIINI environment variable.")
    (println "\nreport bugs to opensource@mnemonic.no")
    (System/exit 0)))

(defn -main
  [& args]
  (let [cli-args (parse-opts args cli-options)
        options (:options cli-args)
        errors (:errors cli-args)]
    (if errors
      (exit (clojure.string/join ", " errors) 1)
      (do
        (when (:help options)
          (usage))
        (System/setProperty "*scio-api-ini*" (:config options))
        (let [ini (clojure-ini/read-ini (get-config-file) :keywordize? true)]
          (run-jetty api
                     {:port (Integer. (get-in ini [:api :port] 3000))}))))))

